src/log/main.cr

```diff
@@ -0,0 +1,130 @@
+class Log
+  # Creates a `Log` for the given source.
+  # If *level* is given, it will override the configuration.
+  def self.for(source : String, level : Severity? = nil) : Log
+    log = builder.for(source)
+    log.level = level if level
+    log
+  end
+
+  # Creates a `Log` for the given nested source.
+  # If *level* is given, it will override the configuration.
+  def for(child_source : String, level : Severity? = nil) : Log
+    log = Log.builder.for(child_source.blank? ? source : "#{source}.#{child_source}")
+    log.level = level if level
+    log
+  end
+
+  private Top = Log.for("")
+
+  {% for method in %i(debug verbose info warn error fatal) %}
+    # See `Log#{{method.id}}`.
+    def self.{{method.id}}(*, exception : Exception? = nil)
+      Top.{{method.id}}(exception: exception) do
+        yield
+      end
+    end
+  {% end %}
+
+  @@builder : Builder?
+
+  # :nodoc:
+  def self.builder
+    @@builder ||= begin
+      builder = Builder.new
+      Log.setup_impl(builder)
+      builder
+    end
+  end
+
+  # :nodoc:
+  def self.setup_impl(config : Configuration)
+  end
+
+  # Use `Log.setup` macro to customize the logging config.
+  # The yielded value *config* is a `Log::Configuration`.
+  # The block should return a truthy value to indicate that the configuration
+  # has been done and there is no need to fallback to other setup methods.
+  #
+  # ```
+  # Log.setup do |config|
+  #   {{config}}.add_backend("stdio", "stdio", Log::Backend::Config.new)
+  #   {{config}}.bind("*", :debug, "stdio")
+  #   true
+  # end
+  # ```
+  #
+  macro setup
+    class ::Log
+      def self.setup_impl(config : Configuration)
+        done = {{yield :config.id}}
```

Comment:
You can introspect the variable name of the first block arg, declare that var, and then paste the block body. Then no `{{config}}` required.

Reply:
👻 but is a pattern it might work for another scenario.